torchvision/prototype/datasets/_builtin/voc.py

```diff
@@ -0,0 +1,141 @@
+        path = pathlib.Path(data[0])
+        return name in path.parent.parts[-depth:]
+
+    def _classify_archive(self, data: Tuple[str, Any], *, config: DatasetConfig) -> Optional[int]:
+        if self._is_in_folder(data, name="ImageSets", depth=2):
+            return 0
+        elif self._is_in_folder(data, name="JPEGImages"):
+            return 1
+        elif self._is_in_folder(data, name=self._ANNS_FOLDER[config.task]):
+            return 2
+        else:
+            return None
+
+    def _decode_detection_ann(self, buffer: io.IOBase) -> torch.Tensor:
+        result = VOCDetection.parse_voc_xml(ElementTree.parse(buffer).getroot())  # type: ignore[arg-type]
+        objects = result["annotation"]["object"]
+        bboxes = [obj["bndbox"] for obj in objects]
+        bboxes = [[int(bbox[part]) for part in ("xmin", "ymin", "xmax", "ymax")] for bbox in bboxes]
+        return torch.tensor(bboxes)
+
+    def _collate_and_decode_sample(
+        self, data, *, config: DatasetConfig, decoder: Optional[Callable[[io.IOBase], torch.Tensor]]
+    ) -> Dict[str, Any]:
+        split_and_image_data, ann_data = data
+        _, image_data = split_and_image_data
+        image_path, image_buffer = image_data
+        ann_path, ann_buffer = ann_data
+
+        image = decoder(image_buffer) if decoder else image_buffer
+
+        if config.task == "detection":
+            ann = self._decode_detection_ann(ann_buffer)
+        else:  # config.task == "segmentation":
+            ann = decoder(ann_buffer) if decoder else ann_buffer
+
+        return dict(image_path=image_path, image=image, ann_path=ann_path, ann=ann)
+
+    def _make_datapipe(
+        self,
+        resource_dps: List[IterDataPipe],
+        *,
+        config: DatasetConfig,
+        decoder: Optional[Callable[[io.IOBase], torch.Tensor]],
+    ) -> IterDataPipe[Dict[str, Any]]:
+        archive_dp = resource_dps[0]
+        archive_dp = TarArchiveReader(archive_dp)
+        split_dp, images_dp, anns_dp = Demultiplexer(
+            archive_dp,
+            3,
+            functools.partial(self._classify_archive, config=config),  # type: ignore[arg-type]
+            drop_none=True,
+            buffer_size=INFINITE_BUFFER_SIZE,
+        )
+
+        split_dp = Filter(split_dp, self._is_in_folder, fn_kwargs=dict(name=self._SPLIT_FOLDER[config.task]))
+        split_dp = Filter(split_dp, path_comparator("name", f"{config.split}.txt"))
+        # TODO: replace the .map() call with decode=True in LineReader when
+        #  https://github.com/pytorch/data/issues/28 is resolved
+        split_dp = LineReader(split_dp).map(bytes.decode, input_col=1)
+        split_dp = Shuffler(split_dp, buffer_size=INFINITE_BUFFER_SIZE)
+
+        dp = split_dp
+        for level, data_dp in enumerate((images_dp, anns_dp)):
+            dp = KeyZipper(
+                dp,
+                data_dp,
+                key_fn=getitem(*[0] * level, 1),
```

Comment:
It seems there is another thing we could improve for `KeyZipper` is how to combine the value together. Currently, the result is`(split, image), ann`

IMO, Not sure if this is possible if it's better or doable to become `split, [image, ann]`.

Reply:
I have the same preference between these two approaches. Then only concern is that buffer is going to blow up if a DataPipe has the reverse order compared to the other DataPipe (buffer becomes M x N, where M represents the number of DataPipes and N represents the number of element). If we stack multiple `KeyZipper`, the same scenario as above that a single DataPipe has the reverse order would cause buffer size N rather than M x N.

Reply:
Zipping multiple datapipes shouldn't be the norm, so I feel like a warning in the documentation should be sufficient.